File: src/services/api/workflows.ts
import { api } from './config';

export type TriggerType = 'email' | 'folder' | 'webhook' | 'manual';
export type ActionType = 'process_document' | 'send_to_erp' | 'send_webhook' | 'send_email' | 'save_to_database' | 'notify_user';

export interface WorkflowTrigger {
  type: TriggerType;
  config: Record<string, any>;
  conditions?: Record<string, any>;
}

export interface WorkflowAction {
  type: ActionType;
  config: Record<string, any>;
  conditions?: Record<string, any>;
  delay_seconds?: number;
}

export interface WorkflowCreate {
  name: string;
  description?: string;
  trigger: WorkflowTrigger;
  actions: WorkflowAction[];
  enabled?: boolean;
  user_id: string;
  organization_id: string;
}

export interface Workflow {
  id: string;
  name: string;
  description?: string;
  trigger: WorkflowTrigger;
  actions: WorkflowAction[];
  enabled: boolean;
  status: string;
  created_at: string;
  updated_at: string;
  execution_count: number;
  success_count: number;
  error_count: number;
}

export async function createWorkflow(workflow: WorkflowCreate) {
  const res = await api.post('/workflows/workflows', workflow);
  return res.data;
}

export async function listWorkflows(userId: string, limit = 20, offset = 0) {
  const res = await api.get('/workflows/workflows', { 
    params: { user_id: userId, limit, offset } 
  });
  return res.data as {
    total: number;
    limit: number;
    offset: number;
    workflows: Workflow[];
  };
}

export async function getWorkflow(workflowId: string, userId: string) {
  const res = await api.get(`/workflows/workflows/${workflowId}`, {
    params: { user_id: userId }
  });
  return res.data as Workflow;
}

export async function updateWorkflow(workflowId: string, updates: Partial<WorkflowCreate>, userId: string) {
  const res = await api.put(`/workflows/workflows/${workflowId}`, updates, {
    params: { user_id: userId }
  });
  return res.data;
}

export async function deleteWorkflow(workflowId: string, userId: string) {
  const res = await api.delete(`/workflows/workflows/${workflowId}`, {
    params: { user_id: userId }
  });
  return res.data;
}

export async function executeWorkflow(workflowId: string, triggerData: Record<string, any>) {
  const res = await api.post(`/workflows/workflows/${workflowId}/execute`, triggerData);
  return res.data;
}

export async function listWorkflowExecutions(workflowId: string, userId: string, limit = 20, offset = 0) {
  const res = await api.get(`/workflows/workflows/${workflowId}/executions`, {
    params: { user_id: userId, limit, offset }
  });
  return res.data as {
    total: number;
    limit: number;
    offset: number;
    executions: any[];
  };
}

File: xagent/xagent/backend/routers/workflows.py
from fastapi import APIRouter, Depends, HTTPException
from typing import List, Dict, Any
from app.auth import verify_token
from pydantic import BaseModel
import uuid
import logging

logger = logging.getLogger(__name__)

router = APIRouter()

# Workflow schemas
class WorkflowStep(BaseModel):
    id: str
    type: str  # 'agent', 'condition', 'data', 'notification'
    action: str
    parameters: Dict[str, Any] = {}
    next_steps: List[str] = []

class WorkflowCreate(BaseModel):
    name: str
    description: str
    steps: List[WorkflowStep]
    trigger_type: str = "manual"  # 'manual', 'scheduled', 'event'
    trigger_config: Dict[str, Any] = {}

class WorkflowUpdate(BaseModel):
    name: str = None
    description: str = None
    steps: List[WorkflowStep] = None
    trigger_type: str = None
    trigger_config: Dict[str, Any] = None

class Workflow(BaseModel):
    id: str
    name: str
    description: str
    steps: List[WorkflowStep]
    trigger_type: str
    trigger_config: Dict[str, Any]
    status: str = "draft"  # 'draft', 'active', 'paused', 'completed'
    created_at: str
    updated_at: str

class WorkflowExecution(BaseModel):
    id: str
    workflow_id: str
    status: str = "running"  # 'running', 'completed', 'failed', 'paused'
    started_at: str
    completed_at: str = None
    current_step: str = None
    results: Dict[str, Any] = {}

# In-memory storage (in production, use database)
workflows_db = []
workflow_executions_db = []

@router.get("/", response_model=List[Workflow])
async def get_workflows(user_id: str = Depends(verify_token)):
    """Get all workflows"""
    return workflows_db

@router.post("/", response_model=Workflow)
async def create_workflow(workflow: WorkflowCreate, user_id: str = Depends(verify_token)):
    """Create a new workflow"""
    try:
        new_workflow = Workflow(
            id=str(uuid.uuid4()),
            name=workflow.name,
            description=workflow.description,
            steps=workflow.steps,
            trigger_type=workflow.trigger_type,
            trigger_config=workflow.trigger_config,
            status="draft",
            created_at=str(uuid.uuid4()),  # In production, use datetime
            updated_at=str(uuid.uuid4())
        )
        workflows_db.append(new_workflow)
        logger.info(f"Created workflow: {new_workflow.name}")
        return new_workflow
    except Exception as e:
        logger.error(f"Failed to create workflow: {e}")
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/{workflow_id}", response_model=Workflow)
async def get_workflow(workflow_id: str, user_id: str = Depends(verify_token)):
    """Get workflow by ID"""
    for workflow in workflows_db:
        if workflow.id == workflow_id:
            return workflow
    raise HTTPException(status_code=404, detail="Workflow not found")

@router.put("/{workflow_id}", response_model=Workflow)
async def update_workflow(
    workflow_id: str, 
    workflow_update: WorkflowUpdate,
    user_id: str = Depends(verify_token)
):
    """Update workflow"""
    for i, workflow in enumerate(workflows_db):
        if workflow.id == workflow_id:
            # Update fields if provided
            if workflow_update.name is not None:
                workflow.name = workflow_update.name
            if workflow_update.description is not None:
                workflow.description = workflow_update.description
            if workflow_update.steps is not None:
                workflow.steps = workflow_update.steps
            if workflow_update.trigger_type is not None:
                workflow.trigger_type = workflow_update.trigger_type
            if workflow_update.trigger_config is not None:
                workflow.trigger_config = workflow_update.trigger_config
            
            workflow.updated_at = str(uuid.uuid4())  # In production, use datetime
            workflows_db[i] = workflow
            return workflow
    
    raise HTTPException(status_code=404, detail="Workflow not found")

@router.delete("/{workflow_id}")
async def delete_workflow(workflow_id: str, user_id: str = Depends(verify_token)):
    """Delete workflow"""
    for i, workflow in enumerate(workflows_db):
        if workflow.id == workflow_id:
            del workflows_db[i]
            return {"message": "Workflow deleted successfully"}
    
    raise HTTPException(status_code=404, detail="Workflow not found")

@router.post("/{workflow_id}/execute")
async def execute_workflow(workflow_id: str, user_id: str = Depends(verify_token)):
    """Execute a workflow"""
    # Find workflow
    workflow = None
    for wf in workflows_db:
        if wf.id == workflow_id:
            workflow = wf
            break
    
    if not workflow:
        raise HTTPException(status_code=404, detail="Workflow not found")
    
    # Create execution record
    execution = WorkflowExecution(
        id=str(uuid.uuid4()),
        workflow_id=workflow_id,
        status="running",
        started_at=str(uuid.uuid4()),
        current_step=workflow.steps[0].id if workflow.steps else None,
        results={}
    )
    
    workflow_executions_db.append(execution)
    
    # Simulate workflow execution
    try:
        for step in workflow.steps:
            execution.current_step = step.id
            
            # Simulate step execution
            result = {
                "step_id": step.id,
                "type": step.type,
                "action": step.action,
                "status": "completed",
                "result": f"Executed {step.action} successfully"
            }
            
            execution.results[step.id] = result
            
            # Simulate some processing time
            import time
            time.sleep(0.1)
        
        execution.status = "completed"
        execution.completed_at = str(uuid.uuid4())
        execution.current_step = None
        
        logger.info(f"Workflow execution completed: {execution.id}")
        return {"execution_id": execution.id, "status": "completed"}
        
    except Exception as e:
        execution.status = "failed"
        execution.completed_at = str(uuid.uuid4())
        logger.error(f"Workflow execution failed: {e}")
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/{workflow_id}/executions", response_model=List[WorkflowExecution])
async def get_workflow_executions(workflow_id: str, user_id: str = Depends(verify_token)):
    """Get executions for a workflow"""
    executions = [exec for exec in workflow_executions_db if exec.workflow_id == workflow_id]
    return executions

@router.get("/executions/{execution_id}", response_model=WorkflowExecution)
async def get_workflow_execution(execution_id: str, user_id: str = Depends(verify_token)):
    """Get workflow execution by ID"""
    for execution in workflow_executions_db:
        if execution.id == execution_id:
            return execution
    raise HTTPException(status_code=404, detail="Workflow execution not found")

@router.post("/{workflow_id}/pause")
async def pause_workflow(workflow_id: str, user_id: str = Depends(verify_token)):
    """Pause a workflow"""
    for workflow in workflows_db:
        if workflow.id == workflow_id:
            workflow.status = "paused"
            logger.info(f"Workflow paused: {workflow_id}")
            return {"message": "Workflow paused successfully"}
    
    raise HTTPException(status_code=404, detail="Workflow not found")

@router.post("/{workflow_id}/resume")
async def resume_workflow(workflow_id: str, user_id: str = Depends(verify_token)):
    """Resume a paused workflow"""
    for workflow in workflows_db:
        if workflow.id == workflow_id:
            workflow.status = "active"
            logger.info(f"Workflow resumed: {workflow_id}")
            return {"message": "Workflow resumed successfully"}
    
    raise HTTPException(status_code=404, detail="Workflow not found")
